native/src/main/java/io/ballerina/stdlib/xmldata/MapFromXml.java

```diff
@@ -0,0 +1,148 @@
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.stdlib.xmldata;
+
+import io.ballerina.runtime.api.TypeTags;
+import io.ballerina.runtime.api.creators.TypeCreator;
+import io.ballerina.runtime.api.creators.ValueCreator;
+import io.ballerina.runtime.api.types.ArrayType;
+import io.ballerina.runtime.api.types.MapType;
+import io.ballerina.runtime.api.types.TableType;
+import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.values.BError;
+import io.ballerina.runtime.api.values.BMap;
+import io.ballerina.runtime.api.values.BString;
+import io.ballerina.runtime.api.values.BTable;
+import io.ballerina.runtime.api.values.BTypedesc;
+import io.ballerina.runtime.api.values.BXml;
+import io.ballerina.runtime.api.values.BXmlSequence;
+import io.ballerina.stdlib.xmldata.utils.Constants;
+import io.ballerina.stdlib.xmldata.utils.XmlDataUtils;
+import org.ballerinalang.langlib.value.CloneWithType;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class converts an XML to a Ballerina record type.
+ *
+ * @since 2.3.0
+ */
+public class MapFromXml {
+
+
+
+    @SuppressWarnings("unchecked")
+    public static Object fromXml(BXml xml, BTypedesc type) {
+        Type describingType = type.getDescribingType();
+        if (describingType.getTag() == TypeTags.RECORD_TYPE_TAG) {
+            try {
+                if (describingType.getFlags() != Constants.DEFAULT_TYPE_FLAG) {
+                    String recordName = describingType.getName();
+                    String elementName = getKey(xml);
+                    if (!recordName.equals(elementName)) {
+                        return XmlDataUtils.getError("The record type name: " + recordName +
+                                " mismatch with given XML name: " + elementName);
+                    }
+                }
+                Object output =  XmlToRecord.covertToJson(xml, true, "", type);
+                if (output instanceof BError) {
+                    return XmlDataUtils.getError("XML type mismatch with record type: " +
+                            ((BError) output).getErrorMessage());
+                }
+                BMap<BString, Object> record = (BMap<BString, Object>) output;
+                if (describingType.getFlags() == Constants.DEFAULT_TYPE_FLAG) {
+                    output = CloneWithType.cloneWithType(output, type);
+                } else {
+                    output = CloneWithType.cloneWithType(record.get(record.getKeys()[0]), type);
```

Comment:
what are we getting from calling `record.getKeys()[0]`. is it guaranteed to return same value always?

Reply:
If the returned type is a user-defined record type, we have to skip the XML parent element and process that child only. So, this is used to get a child.